File: backend/src/main/java/com/laptopdb/backend/entity/Cpu.java
package com.laptopdb.backend.entity;

import java.math.BigDecimal;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Data;

@Data 
@Entity 
@Table(name = "cpus") 
public class Cpu {

    @Id 
    @GeneratedValue(strategy = GenerationType.IDENTITY) 
    private Integer id;

    @NotBlank(message = "Brand cannot be blank")
    @Column(nullable = false, length = 50)
    private String brand;

    @NotBlank(message = "Series cannot be blank")
    @Column(nullable = false, length = 50)
    private String series;

    @NotBlank(message = "Model name cannot be blank")
    @Column(name = "model_name", nullable = false, length = 100, unique = true) 
    private String modelName;

    @NotNull(message = "Base clock frequency cannot be null")
    @Column(name = "base_clock_ghz", nullable = false, precision = 3, scale = 2) 
    private BigDecimal baseClockGhz;

    @Column(name = "boost_clock_ghz", precision = 3, scale = 2)
    private BigDecimal boostClockGhz;

    @NotNull(message = "Core count cannot be null")
    @Column(name = "core_count", nullable = false)
    private Integer coreCount;

    @NotNull(message = "Thread count cannot be null")
    @Column(name = "thread_count", nullable = false)
    private Integer threadCount;

    @NotNull(message = "Cache size cannot be null")
    @Column(name = "cache_mb", nullable = false)
    private Integer cacheMb;
}

File: frontend/lib/models/cpu.dart
class Cpu {
  final int id;
  final String brand;
  final String series;
  final String modelName;
  final double baseClockGhz;
  final double? boostClockGhz;
  final int coreCount;
  final int threadCount;
  final int cacheMb;

  Cpu({
    required this.id,
    required this.brand,
    required this.series,
    required this.modelName,
    required this.baseClockGhz,
    this.boostClockGhz,
    required this.coreCount,
    required this.threadCount,
    required this.cacheMb,
  });

  factory Cpu.fromJson(Map<String, dynamic> json) {
    return Cpu(
      id: json['id'] ?? 0,
      brand: json['brand'] ?? '',
      series: json['series'] ?? '',
      modelName: json['modelName'] ?? '',
      baseClockGhz: (json['baseClockGhz'] ?? 0).toDouble(),
      boostClockGhz: json['boostClockGhz'] != null
          ? (json['boostClockGhz']).toDouble()
          : null,
      coreCount: json['coreCount'] ?? 0,
      threadCount: json['threadCount'] ?? 0,
      cacheMb: json['cacheMb'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'brand': brand,
      'series': series,
      'modelName': modelName,
      'baseClockGhz': baseClockGhz,
      'coreCount': coreCount,
      'threadCount': threadCount,
      'cacheMb': cacheMb,
    };
    if (boostClockGhz != null) map['boostClockGhz'] = boostClockGhz;
    return map;
  }

  String get displayLabel => '$brand $series $modelName';
}
